lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,434 @@
+                return false;
+              }
+              const auto iter = structInfo.find(peer);
+              assert(iter != structInfo.end());
+              const auto &containType = iter->second.containBrgemmCallType;
+              if (containType[BrgemmCallType::DISPATCH] ||
+                  containType[BrgemmCallType::TILECFG]) {
+                return false;
+              }
+            }
+            return true;
+          });
+    } else { // brgemmCallType == BrgemmCallType::TILERELEASE
+      expandInvariantScopeWithCond(
+          structInfo, op,
+          [](Operation *op) -> bool {
+            return !llvm::isa<scf::IfOp, scf::IndexSwitchOp>(op);
+          },
+          [](Operation *self, const OpStructInfoMap &structInfo,
+             Operation *current,
+             const DenseSet<Operation *> &peers) -> bool { return true; });
+    }
+  }
+
+  LogicalResult collectBrgemmContextStructInfo(OpStructInfoMap &structInfo) {
+    // First walk to collect basic structure
+    getOperation()->walk<WalkOrder::PreOrder>(
+        [this, &structInfo](Operation *op) {
+          BrgemmCallType brgemmCallType = getBrgemmCallType(op);
+          if (brgemmCallType == BrgemmCallType::INAPPLICABLE) {
+            return;
+          }
+
+          // Construct the structInfo tree lazily upon encountering BrgemmCall
+          // Op
+          auto info = getOrCreateBrgemmContext(structInfo, op);
+          structInfo.insert(std::make_pair(op, std::move(info)));
+          if (brgemmCallType == BrgemmCallType::TILECFG) {
+            // Also contruct tree node for the input of BrgemmTilecfg for
+            // dependency check in `expandInvariantScope`
+            auto dependOp = op->getOperand(0).getDefiningOp();
+            auto dependInfo = getOrCreateBrgemmContext(structInfo, dependOp);
+            structInfo.insert(std::make_pair(dependOp, std::move(dependInfo)));
+          }
+        });
+
+    // Second walk to analyse hoist related info
+    getOperation()->walk<WalkOrder::PreOrder>(
+        [this, &structInfo](Operation *op) {
+          BrgemmCallType brgemmCallType = getBrgemmCallType(op);
+          if (brgemmCallType != BrgemmCallType::TILECFG &&
+              brgemmCallType != BrgemmCallType::TILERELEASE) {
+            return;
+          }
+
+          // find the maximal invariant scope for hoisting
+          expandInvariantScope(structInfo, op);
+        });
+
+    return success();
+  }
+
+public:
+  using impl::MicrokernelInvariantCodeMotionBase<
+      MicrokernelInvariantCodeMotion>::MicrokernelInvariantCodeMotionBase;
+  void runOnOperation() final {
+    OpStructInfoMap structInfo;
+
+    if (failed(collectBrgemmContextStructInfo(structInfo))) {
+      signalPassFailure();
+    }
+
+    RewritePatternSet patterns(&getContext());
+    patterns.add<BrgemmTilecfgRewriter>(&getContext(), structInfo);
+    patterns.add<BrgemmTilereleaseRewriter>(&getContext(), structInfo);
```

Comment:
Do we have any mechanism to make sure the pair usage of tile cfg and tile release?

Reply:
Actually we don't need to ensure pair usage, we only need to call release once at the end of context. In this pass it just simply remove all release and add one at the end of parallel scope or function.